### PR TITLE
Oculus Mobile SDK WIP

### PR DESCRIPTION
**Pull request by modulesio**
This is a work-in-progress for bootstrapping Android builds.

### CODE REVIEW

**Comment by chrislatorres** [comment]
From @ngokevin :
> Rounded up some information here (sysroot dir, binding.gyp, clang command): https://gist.github.com/ngokevin/7bbcb47bf8bc6bac577dc7d1e26e0eae

File: scripts/build-android.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+pushd ..
+export TARGET_ARCH="aarch64"  # for webrtc install prebuilt
+
+if test -z $NO_INSTALL
+then
+ANDROID=1 npm install --no-optional --verbose --ignore-scripts --target_arch=aarch64
+node scripts/preinstall.js
+fi
+
+OS=android node-gyp rebuild --arch=aarch64 --devdir="$(pwd)/.node-gyp" -v
```
Comment (chrislatorres) [inline scripts/build-android.sh]: `node-gyp` will be required globally here

**Comment by ngokevin** [comment]
Some other information on the PR:

- Includes node-native-*-deps for Android, some copied from ML temporarily until we get the Android build actually linking.

**Comment by chrislatorres** [comment]
The toolchain.zip from `node_modules/android-toolchain/toolchain.zip` is not being seen as a valid archive/not unzipping and is only 134 bytes. 

Whereas the toolchain.zip downloaded directly from [here](https://github.com/ngokevin/android-toolchain/blob/master/toolchain.zip) is 227.3mb. Manually unzipping the contents of toolchain.zip into `node_modules/android-toolchain/`fixed the missing libEGL.so error.

**Comment by ngokevin** [comment]
Hmm, I think it needs Git LFS installed. My local has the toolchain unzipped okay. But it works for you? What errors show up now Maybe it's because I'm on OS X.

**Comment by chrislatorres** [comment]
The setup I am running this on:
Ubuntu 18.04
Node v11.10.1
npm 6.7.0
node-gyp v3.8.0

--------------------------------- 

After running `scripts/build-android.sh` then going into and manually running postinstall:

```
ngokevin/exokit/node_modules/android-toolchain$ npm run postinstall

> android-toolchain@0.0.1 postinstall /home/chris/github/ngokevin/exokit/node_modules/android-toolchain
> node postinstall.js && shx chmod bin/*

Unhandled rejection Error: invalid signature: 0x73726576
    at /home/chris/github/ngokevin/exokit/node_modules/unzipper/lib/parse.js:62:26
    at tryCatcher (/home/chris/github/ngokevin/exokit/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/chris/github/ngokevin/exokit/node_modules/bluebird/js/release/promise.js:510:31)
    at Promise._settlePromise (/home/chris/github/ngokevin/exokit/node_modules/bluebird/js/release/promise.js:567:18)
    at Promise._settlePromise0 (/home/chris/github/ngokevin/exokit/node_modules/bluebird/js/release/promise.js:612:10)
    at Promise._settlePromises (/home/chris/github/ngokevin/exokit/node_modules/bluebird/js/release/promise.js:691:18)
    at Async._drainQueue (/home/chris/github/ngokevin/exokit/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/home/chris/github/ngokevin/exokit/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/home/chris/github/ngokevin/exokit/node_modules/bluebird/js/release/async.js:17:14)
    at processImmediate (timers.js:637:19)

chmod: You must specify a file.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! android-toolchain@0.0.1 postinstall: `node postinstall.js && shx chmod bin/*`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the android-toolchain@0.0.1 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/chris/.npm/_logs/2019-03-08T18_15_39_713Z-debug.log
```

--------------------------------- 

Listing files:
```
ngokevin/exokit/node_modules/android-toolchain$ ls -al
total 40
drwxr-xr-x   2 chris chris  4096 Mar  8 12:17 .
drwxr-xr-x 576 chris chris 20480 Mar  8 12:13 ..
-rw-r--r--   1 chris chris    42 Oct 26  1985 .gitattributes
-rw-r--r--   1 chris chris  1404 Mar  8 12:13 package.json
-rw-r--r--   1 chris chris   462 Oct 26  1985 postinstall.js
-rw-r--r--   1 chris chris   134 Oct 26  1985 toolchain.zip
```

--------------------------------- 

Using unzip:
```
ngokevin/exokit/node_modules/android-toolchain$ unzip toolchain.zip 
Archive:  toolchain.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of toolchain.zip or
        toolchain.zip.zip, and cannot find toolchain.zip.ZIP, period.
```

**Comment by chrislatorres** [comment]
Installing git lfs using [these instructions](https://github.com/git-lfs/git-lfs/wiki/Installation#ubuntu) did not affect the above^

**Comment by chrislatorres** [comment]
After manually unzipping `toolchain.zip` into `android-toolchain`
```
gyp verb cli [ '/home/chris/.nvm/versions/node/v11.10.1/bin/node',
gyp verb cli   '/home/chris/.nvm/versions/node/v11.10.1/bin/node-gyp',
gyp verb cli   'rebuild',
gyp verb cli   '--arch=aarch64',
gyp verb cli   '--devdir=/home/chris/github/ngokevin/exokit/.node-gyp',
gyp verb cli   '-v' ]
gyp info using node-gyp@3.8.0
gyp info using node@11.10.1 | linux | x64
gyp verb command rebuild []
gyp verb command clean []
gyp verb clean removing "build" directory
gyp verb command configure []
gyp verb check python checking for Python executable "python2" in the PATH
gyp verb `which` succeeded python2 /home/chris/github/ngokevin/exokit/node_modules/android-toolchain/bin/python2
gyp ERR! configure error 
gyp ERR! stack Error: Command failed: /home/chris/github/ngokevin/exokit/node_modules/android-toolchain/bin/python2 -c import sys; print "%s.%s.%s" % sys.version_info[:3];
gyp ERR! stack /home/chris/github/ngokevin/exokit/node_modules/android-toolchain/bin/python2: 5: /home/chris/github/ngokevin/exokit/node_modules/android-toolchain/bin/python2: Syntax error: "(" unexpected
gyp ERR! stack 
gyp ERR! stack     at ChildProcess.exithandler (child_process.js:297:12)
gyp ERR! stack     at ChildProcess.emit (events.js:197:13)
gyp ERR! stack     at maybeClose (internal/child_process.js:984:16)
gyp ERR! stack     at Socket.stream.socket.on (internal/child_process.js:401:11)
gyp ERR! stack     at Socket.emit (events.js:197:13)
gyp ERR! stack     at Pipe._handle.close (net.js:611:12)
gyp ERR! System Linux 4.15.0-46-generic
gyp ERR! command "/home/chris/.nvm/versions/node/v11.10.1/bin/node" "/home/chris/.nvm/versions/node/v11.10.1/bin/node-gyp" "rebuild" "--arch=aarch64" "--devdir=/home/chris/github/ngokevin/exokit/.node-gyp" "-v"
gyp ERR! cwd /home/chris/github/ngokevin/exokit
gyp ERR! node -v v11.10.1
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok 
```

-------------------------------

Then manually running the command:
```
$ /home/chris/github/ngokevin/exokit/node_modules/android-toolchain/bin/python2 -c import sys; print "%s.%s.%s" % sys.version_info[:3];
sys.version_info[:3];
bash: /home/chris/github/ngokevin/exokit/node_modules/android-toolchain/bin/python2: cannot execute binary file: Exec format error
Error: no such file "%s.%s.%s"
Error: no such file "%"
Error: no such file ":sys.version_info[:3]"
```

I think is due to the executable being compiled for a different architecture.

**Comment by ngokevin** [comment]
Hmm, maybe the Android toolchain files aren't portable? I'll set up Linux or Windows later. You mentioned the EGL issue went away, but I don't think the command got to that step.

**Comment by chrislatorres** [comment]
This is the missing libEGL.so step that it gets to with running the build script on the current branch, in this case the toolchain.zip hasn't extracted:
```
gyp verb command build []
gyp verb build type Release
gyp verb architecture aarch64
gyp verb node dev dir /home/chris/github/ngokevin/exokit/.node-gyp/11.10.1
gyp verb `which` succeeded for `make` /usr/bin/make
gyp info spawn make
gyp info spawn args [ 'V=1', 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/home/chris/github/ngokevin/exokit/build'
make: *** No rule to make target '/home/chris/github/ngokevin/exokit/node_modules/android-toolchain/sysroot/usr/lib/libEGL.so', needed by '/home/chris/github/ngokevin/exokit/build/Release/libEGL.so'.  Stop.
make: Leaving directory '/home/chris/github/ngokevin/exokit/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/chris/.nvm/versions/node/v11.10.1/lib/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:197:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:254:12)
gyp ERR! System Linux 4.15.0-46-generic
gyp ERR! command "/home/chris/.nvm/versions/node/v11.10.1/bin/node" "/home/chris/.nvm/versions/node/v11.10.1/bin/node-gyp" "rebuild" "--arch=aarch64" "--devdir=/home/chris/github/ngokevin/exokit/.node-gyp" "-v"
gyp ERR! cwd /home/chris/github/ngokevin/exokit
gyp ERR! node -v v11.10.1
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok 
```
full log: https://gist.github.com/chrislatorres/5495966c8b3e35548ebf9603de3c7fc4

**Comment by chrislatorres** [comment]
Building a [standalone toolchain](https://developer.android.com/ndk/guides/standalone_toolchain) fixed the above issue with the exectuables:
```
android-ndk-r19b-linux-x86_64/android-ndk-r19b$ ./build/tools/make_standalone_toolchain.py --arch arm64 --api 21 --install-dir ~/github/test-toolchain
```

**Comment by modulesio** [comment]
I don't think the above actually builds but rather downloads a toolchain. However, to that end, can we integrate it into the build process somehow?

**Comment by ngokevin** [comment]
The EGL I think that's at the GYP preparation stage since I was trying to add flags with the path to try to get it to work. The EGL issues I was facing was during link.

Yeah, I tried to get it so you just have to download it without having to set up NDK and everything. Will see if it works later.

**Comment by chrislatorres** [comment]
note for when building on linux, might run into: https://github.com/webmixedreality/exokit/pull/815

**Comment by chrislatorres** [comment]
Just built successfully on Ubuntu, forked repo here https://github.com/chrislatorres/exokit/commits/android4

The only thing not tracked here is manually building a standalone toolchain and dropping it into `node_modules/android-toolchain/`

--------------------------------------------------

`android-ndk-r19b-linux-x86_64/android-ndk-r19b$ ./build/tools/make_standalone_toolchain.py --arch arm64 --api 21 --install-dir ~/github/test-toolchain`

**Comment by ngokevin** [comment]
Thanks! I'll try it out, and see if I can get the downloadable toolchain working.

File: android/local.properties [hunk 2]
```diff
@@ -0,0 +1,9 @@
+## This file must *NOT* be checked into Version Control Systems,
+# as it contains information specific to your local configuration.
+#
+# Location of the SDK. This is only used by Gradle.
+# For customization when using a Version Control System, please read the
+# header note.
+#Fri Mar 15 18:40:56 CDT 2019
+ndk.dir=/home/chris/Android/Sdk/ndk-bundle
```
Comment (modulesio) [inline android/local.properties]: Is there a way to automate this path?

Comment (chrislatorres) [reply android/local.properties]: `$ANDROID_HOME` in this context points to `/home/chris/Android/Sdk`

so `$ANDROID_HOME/ndk-bundle`

Comment (chrislatorres) [reply android/local.properties]: added 18869398

**Comment by chrislatorres** [comment]
this PR should probably be more specific "Oculus Mobile SDK android support WIP"

**Comment by chrislatorres** [comment]
added the `native-oculus-mobile-deps`:

https://github.com/chrislatorres/native-oculus-mobile-deps 

https://www.npmjs.com/package/native-oculus-mobile-deps

**Comment by chrislatorres** [comment]
Starting work on initial bindings at https://github.com/ngokevin/exokit/pull/6/files

**Comment by modulesio** [comment]
Note the threading comment in https://github.com/webmixedreality/exokit/pull/787/files#r267752453. This doesn't matter for a V1 to ship, but it should be a design consideration that the wait/submit methods should be easy to move to another thread, if they aren't run in a side thread from the start.

**Comment by modulesio** [comment]
Can we document here what the bootstrap procedure for Android Studio is? Something similar to `build-ml.sh` that can be pointed to as an entry point?

My experience with AS has been that if you don't load the project from the right directory, it just locks up.

**Comment by chrislatorres** [comment]
#### Here is a summary of where the progress is, since it is kind of fragmented  



----------------------------------- 

Libnode.a android fs hijack

- https://github.com/webmixedreality/node-magicleap/pull/6 
- Dlib hack
  - https://github.com/webmixedreality/exokit/blob/master/main.cpp#L17-L24 
  - https://github.com/webmixedreality/exokit/blob/master/main.cpp#L173 
- Example dlibs.h
  - https://gist.github.com/chrislatorres/33b33233c6ac1346a3fcd135300bd95a 

----------------------------------- 

# **Native Bindings:**
> “VrApi provides the minimum required API for rendering scenes in VR. Applications may query VrApi for orientation data, and submit textures to apply distortion, sensor fusion, and compositing.
> We have provided the source for VrCubeWorld_NativeActivity and VrCubeWorld_SurfaceView, simple sample applications using VrApi, to serve as references.”

https://developer.oculus.com/documentation/mobilesdk/latest/concepts/mobile-vrapi/#mobile-vrapi 

----------------------------------- 
  

Initial oculus mobile sdk bindings (mostly copy&pasted from https://github.com/webmixedreality/exokit/pull/787 )
- https://github.com/ngokevin/exokit/pull/6 

-----------------------------------

VrApi nativeactivity sample code / native bindings
- https://gist.github.com/chrislatorres/9fd44b39e36208613bf7ee4c032dab9b 


VrApi equivalents:
- (enter XR) requestpresent - oncreate 
  - https://gist.github.com/chrislatorres/9fd44b39e36208613bf7ee4c032dab9b#file-vrcubeworld_nativeactivity-c-L1058 
- (whats the frame pose/input pose) requestframe - get predicted tracking
  - https://gist.github.com/chrislatorres/9fd44b39e36208613bf7ee4c032dab9b#file-vrcubeworld_nativeactivity-c-L1105 
- submitframe - submitframe - happens in same tick
  - https://gist.github.com/chrislatorres/9fd44b39e36208613bf7ee4c032dab9b#file-vrcubeworld_nativeactivity-c-L1130 
- (exit XR) exit - shutdown ovrrender destroy 
  - https://gist.github.com/chrislatorres/9fd44b39e36208613bf7ee4c032dab9b#file-vrcubeworld_nativeactivity-c-L1133 
- Framebuffer
  - https://gist.github.com/chrislatorres/9fd44b39e36208613bf7ee4c032dab9b#file-vrcubeworld_nativeactivity-c-L700 
- Can use same EGL as magic leap
  - https://github.com/webmixedreality/exokit/blob/master/deps/exokit-bindings/egl/src/egl.cc#L208 


----------------------------------- 

Main function
  
- Oculus mobile sdk nativeactivity
  - https://gist.github.com/chrislatorres/9fd44b39e36208613bf7ee4c032dab9b#file-vrcubeworld_nativeactivity-c-L1012-L1017 
- Magic leap (comparison)
  - https://github.com/webmixedreality/exokit/blob/master/main.cpp#L168-L361 

----------------------------------- 

# **Bootstap steps:**
```sh
git clone https://github.com/ngokevin/exokit 
git checkout android3
./scripts/build-android.sh
```

Android studio:
```
import project from `exokit/android` directory
wait for indexing and loading symbols..
build
Run
Logcat for debugging
```

**Comment by chrislatorres** [comment]
Instructions for building the standalone toolchain:

`./make_standalone_toolchain.py --arch arm64 --api 21 --install-dir /output-dir/`

**Comment by modulesio** [comment]
Forked to https://github.com/webmixedreality/exokit/pull/897.

**Pull request closed without merging**